src/features/animation.ts

```diff
@@ -23,121 +23,131 @@ const MILLISECONDS_PER_SECOND = 1000.0
-              return this[$paused];
-            }
-
-            get currentTime(): number {
-              return (this as any)[$scene].model.animationTime;
-            }
-
-            set currentTime(value: number) {
-              (this as any)[$scene].model.animationTime = value;
-            }
-
-            pause() {
-              if (this[$paused]) {
-                return;
-              }
-
-              this[$paused] = true;
-              this.dispatchEvent(new CustomEvent('pause'));
-            }
-
-            play() {
-              if (this[$paused] && this.availableAnimations.length > 0) {
-                this[$paused] = false;
-
-                if (!(this as any)[$scene].model.hasActiveAnimation) {
-                  this[$changeAnimation]();
-                }
-
-                this.dispatchEvent(new CustomEvent('play'));
-              }
-            }
-
-            [$onModelLoad]() {
-              this[$paused] = true;
-
-              if (this.autoplay) {
-                this[$changeAnimation]();
-                this.play();
-              }
-            }
-
-            [$tick](_time: number, delta: number) {
-              super[$tick](_time, delta);
-
-              if (this[$paused]) {
-                return;
-              }
-
-              const {model} = (this as any)[$scene];
-              model.updateAnimation(delta / MILLISECONDS_PER_SECOND);
-
-              this[$needsRender]();
-            }
-
-            updated(changedProperties: Map<string, any>) {
-              super.updated(changedProperties);
-
-              if (changedProperties.has('autoplay') && this.autoplay) {
-                this.play();
-              }
-
-              if (changedProperties.has('animationName')) {
-                this[$changeAnimation]();
-              }
-            }
-
-            async[$updateSource]() {
-              super[$updateSource]();
-
-              // If we are loading a new model, we need to stop the animation of
-              // the current one (if any is playing). Otherwise, we might lose
-              // the reference to the scene root and running actions start to
-              // throw exceptions and/or behave in unexpected ways:
-              (this as any)[$scene].model.stopAnimation();
-            }
-
-            [$changeAnimation]() {
-              const {model} = (this as any)[$scene];
-
-              model.playAnimation(
-                  this.animationName,
-                  this.animationCrossfadeDuration / MILLISECONDS_PER_SECOND);
-
-              // If we are currently paused, we need to force a render so that
-              // the model updates to the first frame of the new animation
-              if (this[$paused]) {
-                model.updateAnimation(0);
-                this[$needsRender]();
-              }
-            }
-          }
-
-          return AnimationModelViewerElement;
-        };
+export interface AnimationInterface {
```

Comment:
You might be able to avoid explicitly defining the mixin interfaces. TypeScript should infer the correct return type of the mixin based on the class returned from the function if you just leave off the return type:

```ts
const MixinA = <T extends Constructor<HTMLElement>>(Base: T) => {
  return class InnerClass extends Base {
    static aStatic = '';
    aInstance = '';
  }
}
```

If you do need an explicit reference to the interface of that inner mixin class, then it looks like a pattern like this could work:

```ts
// Extract the class defined by the mixin:
type MixinAConstructor = ReturnType<typeof MixinA>;
type MixinAInstance = InstanceType<ReturnType<typeof MixinA>>;
declare const ac: MixinAConstructor;
declare const ai: MixinAInstance;
ac.aStatic;
ai.aInstance;
```

Here's a playground where I've been trying out some mixin patterns:

http://www.typescriptlang.org/play/#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqAfFALxQDeUCEA7lABQB0LAhqgOZIBcUrCIANoBdAJS8CUAL4BuAFBzMiFFACyASwAe6hAEEyUQlAibgEBABMksZWiw58ACQKqAMgFEANhAC254EREDABCrEgQEqJkJBRyUFCoEMDoqAhQmJ5h1gCSCDSoMJlI1iZmltah4ZRx8VAorMDqmHwAysANTQYA5F3ytXy59QiY0OQ9fbVgqHBm2BAWfAAK07NmC2O9NVJy23IA9HtQ7qaorNhQwAAW0Dr5ALQ6ZqgAZmfQcM8X11A+WjpQFggzx06kaiDkoEgaj+engyDs2FwBgASkkUggCOAIHhIRAPtDtHoiPJcQSdLpBu1hqMoJT+CNMZA8KjkqlGdjcfiNITdIF5ICMuxoEp4XxMLxueS4SgMIjUPyIILEulbHx1BKYRT4fSIPIzkxWG0Opg9eoDXTqfJFKrJQhggYjKVzFYbPDZQ48M43F5fP5AiEwhEoAQoqQYjVEqy0oLirS8hACkUSqZnRVA9V+vVGs0AEZG7PdTb9HMWkaFibbXYx6yqECVaBO8pQL0ebx+BDADN1doFnOB-OdDYTPvhUs08Y7BQilS23RgMCedTzAy24IMWfruuB0SiK0HKAAdVBlzg6E7V3USAANGS9PPF8vMPwoIhPCAoDnoOhwgswnwLliN4IDM-4AG6sJ46AQEwEJYrec4Lku6xxkMDJYjiWJcpq95IcSCizjh8xMCOEADiacgEYhRGGj2TTyJRD4WMRZHyNOnaaBqPKEchND0AxSEMLuciaMRgZjvIImsOJwnEdJihJkcKaWLaxhKS6-HLrE8RZp0jYWGR5Y1HpY6GfEHBJKpZQWMcYBwOEywzIqayCZQCRoqkXyXgaDmrPMMjSPIuzHFZtqieELFyMFzqhTRxryFFykwsxtHkQlFihXpEVsVA77kLxikhTCgnyCAYUQNJpVSdqlpyKVJbVSMJVMMZDW6rVzVqTZdkQD5Tl+UAA

Reply:
Thanks, that's an awesome insight. I would love it if we could omit the extra interfaces.

Reply:
Amazing. This totally works. Thanks @aomarks 🎉

Reply:
I ran into trouble trying to use this approach while also extending `HTMLElementTagNameMap`. Here is a repro of my error with your playground implementation:

![image](https://user-images.githubusercontent.com/240083/65068064-bf2a5b00-d93c-11e9-90bc-2fb9f4e06cdb.png)

I don't understand where along the way the type is considered to be self-referential, though...

Reply:
Interesting, not sure what's going on there. Can you reproduce that error in the playground?

Reply:
Hopefully / annoyingly, no! http://www.typescriptlang.org/play/#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqAfFALxQDeUCEA7lABQB0LAhqgOZIBcUrCIANoBdAJS8CUAL4BuAFByIADzC5gUTIhRQAsgEslehAEEyUQlGXAICACZJYWtFhz4AEgR0AZAKIAbCABbG2AiIgYAIVYkCAlRMhIKOSgoVAhgdFQEDT9ohwBJBBpUGFykBysbeygomMpklKgUVmA9TD4AZWAWtrMAcj75Rr5C5oRMaHIBocawVDhrbAhbPgAFecXrFanBhqk5fbkAeiOoHyU0VmwoYAALaCNigFoja1QAMyvoOHeb+6hAgYjFBbBB3kY9K1EIoVGobuBoPpDCZ4MhnNhcGYAErpTIIAgIvCgSA-XRAkxEeTKVSodTExHk4yjbrjSZQZn8CYEyB4HEZLLciBEhGkpFGYxheSgzC5NIaJx8TC8MUopwYDGoKUQGXsaCaNF8PTKxkc1nyK5MVhdHqYc16S2mibyGE09T67QqiJmCyVOwOVEodWuPAebz+IIhMKRaKxKAEeKkRINNL87I68rsooQEplCoXKoOWrQJLDZqtdoAI2t5f6u2GFcdbOmewOCmpcPTDh0ICLlnzfqgod8AWCCHUJaa3RrFZj1d6OxmM5ijdr8kO7dp8oNKuMYDAfj0yzMnoYO9PPZjolEzpOUAA6pDbnB0HTbnokAAaMnI3f7w8rTB+CgRA-BAKAK2gdAYhWaI+HhSAvwQBY4IAN1YPx0AgJgXThelv3FPcDyPchG0FYUSV+HdCP-SkFCov9liYJcIDnW05HoojbEtVj5A4-8mJ4uR3XUJRjR-ajiOoOh8JMCTbAYa85CUJiY0beRlNYNSlKYrScM3Tszn7WwVT7awBz4o8JzLXpfVsVjVwaWyVwXBoOHSUyqnOVQYnWBZtS2BTKFSXEsj+d9LV8zZlhkaQ1wUc4zOM8kVJiQSEqqFVuKnNp5HSuxMqrbK2LypLkSYWzBOEqAwPIGh6BKlUFPkEAUogLSWs0tFOQgZqdK6s05Bapz+qdQbyqMry4B8jZ-OihRpVlaAOD8OAZz8eoUlebNPgmQdPGHCMxwIVgOAAOVYYIdFYMANsaPpAWRJ5roY2w+jEgiXpmfYpCAA

I wonder what the meaningful delta could be...

Reply:
One potentially significant difference is the compiler version. We are using 3.6.3, while the playground is apparently on 3.5.1.

Reply:
Sure enough, 3.5.1 compiles fine. 3.6.3 shows the error.

Reply:
I pinned us to 3.5.3, as this is the "stable" version used internally. We'll save this issue for another day.

Reply:
SGTM!